cosmos/operators/watcher.py

```diff
@@ -90,3 +118,132 @@
+          ``dbt_startup_events``;
+        – pushes each ``NodeFinished`` event immediately to XCom under
+          ``nodefinished_<unique_id>`` (gzipped+base64 JSON) so downstream
+          sensors can react with near-zero latency.
+
+    - **When ``dbtRunner`` is *not* available** (older dbt or
+      ``InvocationMode=SUBPROCESS``) we fallback to delayed strategy: after
+      dbt exits we read ``target/run_results.json`` and push the whole mapping
+      once under key ``run_results`` to XCom.  Sensors can poll this key but will not
+      get per-model updates until the build completes.
+
+    This keeps the heavy dbt work centralised while providing near real-time
+    feedback and granular task-level observability downstream.
+    """
+
+    def __init__(
+        self,
+        *,
+        render_config: RenderConfig | None = None,
+        **kwargs: Any,
+    ) -> None:
+        # Store so we can honour select/exclude when building flags
+        self.render_config: RenderConfig | None = render_config
+
+        task_id = kwargs.pop("task_id", "dbt_build_coordinator")
+        super().__init__(task_id=task_id, **kwargs)
+
+    def add_cmd_flags(self) -> list[str]:
+        flags: list[str] = super().add_cmd_flags()
+
+        self.log.info("DbtBuildCoordinatorOperator: render_config: %s", self.render_config)
+        if self.render_config is not None and self.render_config.exclude:
+            flags.extend(["--exclude", *self.render_config.exclude])
+        if self.render_config is not None and self.render_config.select:
+            flags.extend(["--select", *self.render_config.select])
+        return flags
+
+    @staticmethod
+    def _serialize_event(ev: EventMsg) -> dict[str, Any]:
+        """Convert structured dbt EventMsg to plain dict."""
+        from google.protobuf.json_format import MessageToDict
+
+        return MessageToDict(ev, preserving_proto_field_name=True)  # type: ignore[no-any-return]
+
+    def _handle_startup_event(self, ev: EventMsg, startup_events: list[dict[str, Any]]) -> None:
+        info = ev.info  # type: ignore[attr-defined]
+        raw_ts = getattr(info, "ts", None)
+        ts_val = raw_ts.ToJsonString() if hasattr(raw_ts, "ToJsonString") else str(raw_ts)  # type: ignore[union-attr]
+        startup_events.append({"name": info.name, "msg": info.msg, "ts": ts_val})
+
+    def _handle_node_finished(
+        self,
+        ev: EventMsg,
+        context: Context,
+    ) -> None:
+        self.log.debug("DbtBuildCoordinatorOperator: handling node finished event: %s", ev)
+        ti = context["ti"]
+        uid = ev.data.node_info.unique_id
+        ev_dict = self._serialize_event(ev)
+        payload = base64.b64encode(gzip.compress(json.dumps(ev_dict).encode())).decode()
+        ti.xcom_push(key=f"nodefinished_{uid.replace('.', '__')}", value=payload)
+
+    @contextmanager
+    def _patch_runner(self, callback: Callable[[EventMsg], None]) -> Any:
```

Comment:
Why do we need to patch the runner? A docstring could be very useful

Reply:
removed the patching and added capability to pass callbacks to base get_runner() method